<h6>Fall 2021 - CPSC 483-02<br>
Project 1 - Anomaly Detection<br>
Sean Javiya <br>
Jake Wong <br>
Timothy Jan <br>
</h6>
In Project 1 - Anomaly Detection, we were given a dataset containing a list of students and their Zoom attendance over a 5 week course.<br>We performed various statistical tests on the data, in order to develop a function that will parse the provided dataset and output a list of students who were tardy or under-attended.<br>We developed two functions using different statistical methods of detecting anomalies in the dataset.<br>We conducted an analysis on the results and determined which function was more appropriate for the given dataset.

<h1>Experiment 1: Use the csv module to load the dataset and examine the contents of the first few rows.</h1>

In [1]:
import csv
import statistics
headers = []
students = []

with open('participants.csv') as csv_data:
    data_stream = csv.reader(csv_data, delimiter=',')
    headers = next(data_stream)
    students = [row for row in data_stream]   
    student_names = [names[0] for names in students]
    statistics_data = [int(week1data[1]) for week1data in students]

studentcount = len(students)
weekcount = len(headers) - 1    #the first column is 'Student Name', so we subtract 1 from the length of headers

print(f'There are {studentcount} students in the {weekcount} week course.')
print('The first 5 students and their attendance over the course are as follows:')
print(headers)
for i in range(5):
    print(students[i])

There are 35 students in the 5 week course.
The first 5 students and their attendance over the course are as follows:
['Student Name', 'Week 1', 'Week 2', 'Week 3', 'Week 4', 'Week 5']
['Adrian Ellison', '77', '154', '4', '170', '175']
['Ophelia Mcphee', '179', '151', '164', '173', '171']
['Yasir Fenton', '180', '47', '164', '168', '169']
['Benny Arias', '180', '152', '161', '170', '170']
['Tamara Cottrell', '183', '79', '161', '173', '168']


Here we have imported the csv module to open our csv file "<b>participants.csv</b>" and then we load the dataset into four lists: headers, students, student_names, and statistics_data<br><ol>
<li><b><code>headers</code></b> contains the elements from the first row from the csv file; it is a list of the headers of each column.<br></li>
<li><code><b>students</b></code> is a list of lists.  "students" contains every field from the entire csv except the headers row. It is a list where each element corresponds to an entire row.  Each row element is a list of the values from each column.<br></li>
<li><code><b>student_names</b></code> contains the first column of "students"; it is a list of the names of every student, in the order they appear in the csv file.<br></li>
<li><code><b>statistics_data</b></code> is a list of all the values (time logged in) from a specified column (a specific week of the course).<br></li>

> <i><code>statistics_data</code> is initialized to hold the values from the Week 1 column.</i>


<h1>Experiment 2: Use the statistics module to find the mean and median of Week 1’s data.</h1>

In [2]:
mean = statistics.mean(statistics_data)
median = statistics.median(statistics_data)

print(f'Mean of Week 1 data:   {mean}')
print(f'Median of Week 1 data: {median}')

Mean of Week 1 data:   161
Median of Week 1 data: 175


In <b>Experiment 2</b>, we have used the statistics module to find the <b>mean</b> and <b>median</b> of <code>statistics_data</code>.<br>
<br><b>We have obtained the mean and median of the logged minutes during Week 1.</b><br>
> *Recall that <code>statistics_data</code> was initialized to hold the values from the Week 1 column.*

<h1>Experiment 3: Find the quartiles for Week 1.</h1>

In [3]:
quartiles = statistics.quantiles(statistics_data)

quartiles

[174.0, 175.0, 179.0]

<h3>In <b>Experiment 3</b>, we used the statistics module to find the <b>quartiles</b> of <code>statistics_data</code>.<br></h3>
We save the quartiles in order as:<br>
<ul><li><code>quartiles[0]</code></li><br>
<li><code>quartiles[1]</code></li><br>
<li><code>quartiles[2]</code></li></ul><br>
<b>We have obtained the quartiles of the logged minutes during Week 1.</b><br>

> *Recall that <code>statistics_data</code> was initialized to hold the values from the Week 1 column.*

<br>Please note:<br><code>statistics.quantiles(statistics_data)</code> defaults to n=4 when n is not specified.<br>Therefore, this function call is equivalent to:<br>
<code>statistics.quantiles(statistics_data, n=4)</code>

<h1>Experiment 4: Use the Tukey Fences method to find the outliers in the Week 1 attendance data.</h1>

In [4]:
over_attended = []
under_attended = []
k_factor = 1.5
outlier_threshold = [(quartiles[0] - (k_factor * (quartiles[2] - quartiles[0]))), 
                     (quartiles[2] + (k_factor * (quartiles[2] - quartiles[0])))]

for x in range(studentcount):
    if statistics_data[x] > outlier_threshold[1]:
        to_be_added = (student_names[x], statistics_data[x])
        over_attended.append(to_be_added)
    if statistics_data[x] < outlier_threshold[0]:
        to_be_added = (student_names[x], statistics_data[x])
        under_attended.append(to_be_added)

if over_attended:
    print("The outliers who over attended during Week 1 are:")
    for outlier in over_attended:
        print(f"{outlier[0]}\t\twho spent {outlier[1]} minutes in Zoom.")
if under_attended:
    print("The outliers who under attended during Week 1 are:")
    for outlier in under_attended:
        print(f"{outlier[0]}\t\twho spent {outlier[1]} minutes in Zoom.")
if (not under_attended) and (not over_attended):
    print("There are no outliers.")  

The outliers who under attended during Week 1 are:
Adrian Ellison		who spent 77 minutes in Zoom.
Tayla Sparrow		who spent 51 minutes in Zoom.
Owain Emerson		who spent 9 minutes in Zoom.
Alaya Dickinson		who spent 24 minutes in Zoom.


<h3>In <b>Experiment 4</b>, we find the <b>outliers</b> (of the Week 1 data) using the <b>Tukey Fences</b> method.</h3><br>
Using Tukey Fences is a method to find outliers in terms of the <b>interquartile range</b>.<br><br>
We can find the outliers using the interquartile range. For example, if <code>quartiles[0]</code>(or ${Q_1}$) and <code>quartiles[2]</code>(or $Q_{3}$) are the lower and upper quartiles respectively, then an outlier is any value outside the range:<br>

> $[Q_{1}-k(Q_{3}-Q_{1}), Q_{3}+k(Q_{3}-Q_{1})]$<br>

<code>outlier_threshold = [(quartiles[0] - (k_factor * (quartiles[2] - quartiles[0]))),</code><br>
<code>                     (quartiles[2] + (k_factor * (quartiles[2] - quartiles[0])))]</code>
<br>
for some nonnegative constant <code><b>k_factor</b></code>(or ${k}$). <br>
*John Tukey proposed this test, where k_factor = <b>1.5</b> indicates an "outlier", and k_factor = <b>3</b> indicates data that is "far out"* 
<br><br>
We used the quartiles obtained from Experiment 3 to find the range for the outliers of the Week 1 data using the Tukey fences method.<br>
Finally, we compare the values from the Week 1 data with our upper and lower bounds to find and output the outliers to the standard output device.<br><br>
<b>We have obtained the outliers of the Week 1 data using the Tukey Fences method.</b><br>
> *Recall that <code>statistics_data</code> was initialized to hold the values from the Week 1 column.*

<h1>Experiment 5: Compute the standard deviation for the Week 1 attendance data, then use the empirical rule to find the outliers.</h1>

In [5]:
outside_3_stdev = []

std_dev = statistics.stdev(statistics_data)
lower_bound = mean - (3 * std_dev)
upper_bound = mean + (3 * std_dev)

for x in range(studentcount):
    to_be_compared = statistics_data[x]
    if to_be_compared <  lower_bound or to_be_compared > upper_bound:
        to_be_added = (student_names[x], statistics_data[x])
        outside_3_stdev.append(to_be_added)

print(f"Week 1's Standard Deviation: {std_dev}\n")
if not outside_3_stdev:
    print('There are no outliers.')
else:
    print("The outliers of Week 1 are:")
    for outlier in outside_3_stdev:
        print(f"{outlier[0]}\t\twho spent {outlier[1]} minutes in Zoom.")

Week 1's Standard Deviation: 45.02025687856262

The outliers of Week 1 are:
Owain Emerson		who spent 9 minutes in Zoom.
Alaya Dickinson		who spent 24 minutes in Zoom.


<h3>In <b>Experiment 5</b>, we find the <b>outliers</b> (of the Week 1 data) using the <b>empirical rule</b>.</h3>
<br>
The empirical rule, or the 68–95–99.7 rule, represent the percentage of values that lie within an interval estimate in a normal distribution: According to the empirical rule, 99.7% of the values lie within three standard deviations of the mean.<br>

> $Probability(\mu-3\sigma \leq X \leq \mu + 3\sigma) \approx 99.73\%$<br>

$\mu$ is the Mean of the data<br>
$\sigma$ is the standard deviation from the mean<br>
X is a data point from our normally distributed data<br>
<b>Assuming that our data is normally distributed</b>, this gives us another way to find <b>outliers</b>.<br><br>
We use the statistics module to find the standard deviation of the values in <code>statistics_data</code>.<br>
We save the standard deviation as <code><b>std_dev</b></code><br>

> *Recall that <code>statistics_data</code> was initialized to hold the values from Week 1.*

We use the <b>mean</b> obtained from Experiment 2 to calculate the <b>upper and lower bounds of our distribution</b>.<br>
<code>lower_bound = mean - (3 * std_dev)</code><br>
<code>upper_bound = mean + (3 * std_dev)</code><br>

> *Recall that <code>std_dev</code> is the standard deviation we obtained from the statistics module.*<br>

<br>
<b>Finally, we compare the values from the Week 1 data with our upper and lower bounds to find and output the outliers to the standard output device.</b><br><br>

<u><b><i>The results of Experiment 5 do not agree with the reults of Experiment 4.</i></b></u>

<h1>Experiment 6: Define a function tardy_iqr() to repeat experiment 4.</h1>

In [6]:
def tardy_iqr(week):
    over_attended = []
    under_attended = []
    k_factor = 1.5
    if not headers:
        print("Please select 'Run All Cells'")
        return
    elif week not in headers:
        print(f"{week} was not found within the list of Headers")
        return
    elif week == "Student Name":
        print("'Student Name' is not a valid option")
        return
    week_index = headers.index(week)
    statistics_data = [int(row[week_index]) for row in students]
    quartiles = statistics.quantiles(statistics_data)
    outlier_threshold = [(quartiles[0] - (k_factor * (quartiles[2] - quartiles[0]))), 
                         (quartiles[2] + (k_factor * (quartiles[2] - quartiles[0])))]
    
    for x in range(studentcount):
        if statistics_data[x] > outlier_threshold[1]:
            to_be_added = (student_names[x], statistics_data[x])
            over_attended.append(to_be_added)
        if statistics_data[x] < outlier_threshold[0]:
            to_be_added = (student_names[x], statistics_data[x])
            under_attended.append(to_be_added)
    if not under_attended:
        print(f"There are no students who under attended during {week}.") 
    else:
        print(f"The students who under attended during {week}:")
        for student in under_attended:
            print(f"{student[0]}\t\t\twho spent {student[1]} minutes in Zoom.")
    
    names_of_under_attended = [name[0] for name in under_attended]
    return names_of_under_attended

tardy_iqr("Week 1")

The students who under attended during Week 1:
Adrian Ellison			who spent 77 minutes in Zoom.
Tayla Sparrow			who spent 51 minutes in Zoom.
Owain Emerson			who spent 9 minutes in Zoom.
Alaya Dickinson			who spent 24 minutes in Zoom.


['Adrian Ellison', 'Tayla Sparrow', 'Owain Emerson', 'Alaya Dickinson']

<h3>In <b>Experiment 6</b>, we built a <b>function to perform Experiment 4</b>.<br></h3>
<code>tardy_iqr()</code><br>
The function <code>tardy_iqr()</code> takes one parameter.<br>The <b>parameter</b> is the Week # that we want Experiment 4 performed upon.<br>The function <code>tardy_iqr()</code> <b>returns</b> a list of names (of students who under attended).<br><br>
First, we verify that the given parameter exists in <code><b>headers</b></code>.<br>

> *Recall that <code>headers</code> contains the headers of each column.*

Second, we initialize <b><code>statistics_data</code></b> to contain the values from the specified week (the parameter).<br>We copy the desired values from <code><b>students</b></code> to prevent wasting resources by re-opening our csv file.<br>

> *Recall that <code>students</code> contains every field from the entire csv file except the headers row.*<br>

Third, we perform Experiment 4 and find the students who under attended.<br><br>
<b>Finally, we output a detailed report to the standard output device and return the names of the students who under attended.</b>

<h1>Experiment 7: Define a function tardy_stdev() to repeat experiment 5.</h1> 

In [7]:
def tardy_stdev(week):
    outside_3_stdev = []
    if not headers:
        print('Please select "Run All Cells"')
        return
    if week not in headers:
        print(f"{week} was not found within the list of Headers")
        return
    elif week == "Student Name":
        print("'Student Name' is not a valid option")
        return
    week_index = headers.index(week)
    statistics_data = [int(row[week_index]) for row in students]
    mean = statistics.mean(statistics_data)
    median = statistics.median(statistics_data)
    std_dev = statistics.stdev(statistics_data)
    lower_bound = mean - (3 * std_dev)
    upper_bound = mean + (3 * std_dev)
    print(f'Mean of {week} data:   {mean}')
    print(f'Median of {week} data: {median}')
    print(f"Standard Deviation of {week}: {std_dev}")
    for x in range(studentcount):
        to_be_compared = statistics_data[x]
        if to_be_compared <  lower_bound or to_be_compared > upper_bound:
            to_be_added = (student_names[x], statistics_data[x])
            outside_3_stdev.append(to_be_added)
    if not outside_3_stdev:
        print(f"There are no outliers during {week}.") 
    else:
        print(f'The outliers of {week} are:')
        for student in outside_3_stdev:
            print(f"{student[0]}\t\t\twho spent {student[1]} minutes in Zoom.")

    names_of_outliers = [name[0] for name in outside_3_stdev]
    return names_of_outliers

tardy_stdev("Week 1")

Mean of Week 1 data:   161
Median of Week 1 data: 175
Standard Deviation of Week 1: 45.02025687856262
The outliers of Week 1 are:
Owain Emerson			who spent 9 minutes in Zoom.
Alaya Dickinson			who spent 24 minutes in Zoom.


['Owain Emerson', 'Alaya Dickinson']

<h3>In <b>Experiment 7</b>, we built a <b>function to perform Experiment 5</b>.<br></h3>
<code>tardy_stdev()</code><br>
The function <code>tardy_stdev()</code> takes one parameter.<br>The <b>parameter</b> is the Week # that we want Experiment 5 performed upon.<br>The function <code>tardy_stdev()</code> <b>returns</b> a list of names (of students who were outliers).<br><br>
First, we verify that the parameter exists in <b><code>headers</code></b>.<br>

> *Recall that "headers" contains the headers of each column.*<br>

Second, we initialize <code><b>statistics_data</b></code> to contain the values from the specified week (the parameter).<br>
We take the values from <code><b>students</b></code>" to prevent wasting resources by re-opening our csv file.<br>

> *Recall that <code>students</code> contains every field from the entire csv file except the headers row.*<br>

Third, we perform Experiment 5 and find the students who were outliers.<br><br>
<b>Finally, we output a detailed report to the standard output device and return the names of the students who were outliers.</b>

<h1>Experiment 8: Compare the results of tardy_iqr() and tardy_stdev() on Weeks 2-5.</h1>

In [8]:
x = 2
compare_iqr = []
compare_stdev = []
while x <= 5:
    print(f'                         Week {x}')
    print('--------------------------------------------------------------')
    print(f'Comparing tardy_iqr to tardy_stdev for Week {x}:')
    print('1)     tardy_iqr (using Tukey Fences):')
    compare_iqr = tardy_iqr(f'Week {x}')
    print('\n2)     tardy_stdev (using Standard Deviation):')
    compare_stdev = tardy_stdev(f'Week {x}')
    print('--------------------------------------------------------------')
    print('tardy_iqr returned:')
    print(compare_iqr)
    print('\ntardy_stdev returned:')
    print(compare_stdev)
    print('--------------------------------------------------------------')
    print('--------------------------------------------------------------\n')
    x += 1


                         Week 2
--------------------------------------------------------------
Comparing tardy_iqr to tardy_stdev for Week 2:
1)     tardy_iqr (using Tukey Fences):
The students who under attended during Week 2:
Yasir Fenton			who spent 47 minutes in Zoom.
Tamara Cottrell			who spent 79 minutes in Zoom.
Jazmin Foreman			who spent 103 minutes in Zoom.
Bear Zuniga			who spent 131 minutes in Zoom.
Miles Lyons			who spent 2 minutes in Zoom.

2)     tardy_stdev (using Standard Deviation):
Mean of Week 2 data:   145.0857142857143
Median of Week 2 data: 153
Standard Deviation of Week 2: 41.76712148298934
The outliers of Week 2 are:
Miles Lyons			who spent 2 minutes in Zoom.
Owain Emerson			who spent 290 minutes in Zoom.
--------------------------------------------------------------
tardy_iqr returned:
['Yasir Fenton', 'Tamara Cottrell', 'Jazmin Foreman', 'Bear Zuniga', 'Miles Lyons']

tardy_stdev returned:
['Miles Lyons', 'Owain Emerson']
--------------------------------------

<h3>In <b>Experiment 8</b>, we <b>compare the results of</b> <code>tardy_iqr()</code> and <code>tardy_stdev()</code><b> for Weeks 2 through 5.</b></h3><br>
Most of results from these functions differ.  So we are posed with the question, which method is more correct for this data set?<br><ul>
<li><b><u>Week 2:</u></b></li>
    <code>tardy_iqr()</code> returned 'Yasir Fenton', 'Tamara Cottrell', 'Jazmin Foreman', 'Bear Zuniga' and 'Miles Lyons'.<br>
    <code>tardy_stdev()</code> returned 'Miles Lyons', 'Owain Emerson'.<br>
    <b>We must observe the other data from the report printed to the standard ouptut to make a decision.</b><br>
    Mean of Week 2 data:   145.0857142857143<br>
    Median of Week 2 data: 153<br>
    Standard Deviation of Week 2: 41.76712148298934<br>
    Yasir Fenton			who spent 47 minutes in Zoom.<br>
    Tamara Cottrell			who spent 79 minutes in Zoom.<br>
    Jazmin Foreman			who spent 103 minutes in Zoom.<br>
    Bear Zuniga			who spent 131 minutes in Zoom.<br>
    Miles Lyons			who spent 2 minutes in Zoom.<br>
    Owain Emerson			who spent 290 minutes in Zoom<br>
    <br>
    <b><u>Decision:</u>     <code>tardy_iqr()</code></b><br>
    Yasir, Tamara, Jazmin, Bear, and Miles were all more than 15 minutes below the mean.<br>
    Therefore, <code>tardy_iqr()</code> provided the more accurate results, since <code>tardy_stdev()</code> did not include Yasir, Tamara, Jazmin, or Bear.<br>Furthermore, <code>tardy_stdev()</code> also returned Owain, who actually over attended during Week 2.<br>It might not be correct to mark Owain as tardy, they might have forgot to log off Zoom after class.
    <br><br>
<li><b><u>Week 3:</u></b></li>
    <code>tardy_iqr()</code> returned 'Adrian Ellison', 'Adeline Jordan', 'Jaye Sweeney'.<br>
    <code>tardy_stdev()</code> returned 'Adrian Ellison'.<br>
    <b>We must observe the other data from the report printed to the standard ouptut to make a decision.</b><br>
    Mean of Week 3 data:   153.8<br>
    Median of Week 3 data: 161<br>
    Standard Deviation of Week 3: 28.709293432820054<br>
    Adrian Ellison			who spent 4 minutes in Zoom.<br>
    Adeline Jordan			who spent 105 minutes in Zoom.<br>
    Jaye Sweeney			who spent 121 minutes in Zoom.<br>
    <br>
    <b><u>Decision:</u>     <code>tardy_iqr()</code></b><br>
    Adrian, Adeline, and Jaye were all more than 30 minutes below the mean.<br>
    Therefore, <code>tardy_iqr()</code> provided the more accurate results, since <code>tardy_stdev()</code> did not include Adeline and Jaye.<br>
    <br>
<li><b><u>Week 4:</u></b></li>
    <code>tardy_iqr()</code> returned an empty list.<br>
    <code>tardy_stdev()</code> returned 'Dora Delacruz'.<br>
    <b>We must observe the other data from the report printed to the standard ouptut to make a decision.</b><br>
    Mean of Week 4 data:   175.31428571428572<br>
    Median of Week 4 data: 171<br>
    Standard Deviation of Week 4: 24.63590329361676<br>
    Dora Delacruz			who spent 316 minutes in Zoom<br>
    <br>
    <b><u>Decision:</u>     <code>tardy_iqr()</code></b><br>
    No students logged minutes considerably below the mean.<br>
    Therefore, <code>tardy_iqr()</code> provided the more accurate results.
    <br>Furthermore, <code>tardy_stdev()</code> also returned Dora, who actually over attended during Week 4.
    <br>It might not be correct to mark Dora as tardy, they might have forgot to log off Zoom after class.
    <br><br>
<li><b><u>Week 5:</u></b></li>
    <code>tardy_iqr()</code> returned 'Jazmin Foreman', 'Sanjay Edwards', 'Alfie-James Pierce', 'Saffa Brook'.<br>
    <code>tardy_stdev()</code> returned 'Jazmin Foreman'.<br>
    <b>We must observe the other data from the report printed to the standard ouptut to make a decision.</b><br>
    Mean of Week 5 data:   161.8<br>
    Median of Week 5 data: 171<br>
    Standard Deviation of Week 5: 33.36941184766108<br>
    Jazmin Foreman			who spent 35 minutes in Zoom.<br>
    Sanjay Edwards			who spent 66 minutes in Zoom.<br>
    Alfie-James Pierce	    who spent 74 minutes in Zoom.<br>
    Saffa Brook			    who spent 143 minutes in Zoom.<br>
    <br>
    <b><u>Decision:</u>     <code>tardy_iqr()</code></b><br>
    Jazmin, Sanjay, Alfie-James, and Saffa were all more than 18 minutes below the mean.<br>
    Therefore, <code>tardy_iqr()</code> provided the more accurate results.<br>
    Furthermore, <code>tardy_stdev()</code> did not return Sanjay and Alfie-James, even though they logged around 100 minutes below the mean.
</ul><br>
<b><u>Conclusions:</u></b><br>
For every week, <code>tardy_iqr()</code> was the more appropriate method of finding tardy students.<br>
<code>tardy_stdev()</code> included students who over-attended, and did not include enough students who under-attended.<br>
This might indicate that our data is not normally distributed, as we assumed.  In order for the empirical rule to produce correct results, the values must be normally distributed.<br>
A future area of research would be to use a Q-Q Plot or another method to test if our dataset is normally distributed.